src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -97,169 +135,335 @@ func newCommitLogSource(opts Options, inspection fs.Inspection) bootstrap.Source
+					// Not bootstrapping this namespace.
+					ns = &bootstrapNamespace{
+						namespaceID:   nsIDCopy,
+						bootstrapping: false,
+					}
+				} else {
+					// Bootstrapping this namespace.
+					nsMetadata := nsResult.namespace.Metadata
+					ns = &bootstrapNamespace{
+						namespaceID:             nsIDCopy,
+						bootstrapping:           true,
+						dataAndIndexShardRanges: nsResult.dataAndIndexShardRanges,
+						namespace:               nsMetadata,
+						namespaceContext:        namespace.NewContextFrom(nsMetadata),
+						dataBlockSize:           nsMetadata.Options().RetentionOptions().BlockSize(),
+						accumulator:             nsResult.namespace.DataAccumulator,
+					}
+				}
+				// Append for quick re-lookup with other series.
+				commitLogNamespaces = append(commitLogNamespaces, ns)
+			}
+
+			// Resolve the series in the accumulator.
+			accumulator := ns.accumulator
+
+			// NB(r): Make only encoded tags is used and not series.Tags (we
+			// explicitly ask for references to be returned and to avoid
+			// decoding the tags if we don't have to).
+			if decodedTags := len(entry.Series.Tags.Values()); decodedTags > 0 {
+				msg := "commit log reader expects encoded tags"
+				instrumentOpts := s.opts.ResultOptions().InstrumentOptions()
+				instrument.EmitAndLogInvariantViolation(instrumentOpts, func(l *zap.Logger) {
+					l.Error(msg, zap.Int("decodedTags", decodedTags))
+				})
+				err := instrument.InvariantErrorf(fmt.Sprintf("%s: decoded=%d", msg, decodedTags))
+				return bootstrap.NamespaceResults{}, err
+			}
+
+			var tagIter ident.TagIterator
+			if len(entry.Series.EncodedTags) > 0 {
+				tagDecoderCheckedBytes.Reset(entry.Series.EncodedTags)
+				tagDecoder.Reset(tagDecoderCheckedBytes)
+				tagIter = tagDecoder
+			}
+
+			// Check out the series for writing.
+			series, err := accumulator.CheckoutSeries(entry.Series.ID, tagIter)
+			if err != nil {
+				return bootstrap.NamespaceResults{}, err
+			}
+
+			seriesEntry = seriesMapEntry{
+				namespace: ns,
+				series:    series,
+			}
+			commitLogSeries[seriesKey] = seriesEntry
 		}
 
-		datapointsRead++
-
-		// Distribute work such that each encoder goroutine is responsible for
-		// approximately numShards / numConc shards. This also means that all
-		// datapoints for a given shard/series will be processed in a serialized
-		// manner.
-		// We choose to distribute work by shard instead of series.UniqueIndex
-		// because it means that all accesses to the shardDataByShard slice don't need
-		// to be synchronized because each index belongs to a single shard so it
-		// will only be accessed serially from a single worker routine.
-		encoderChans[series.Shard%uint32(numConc)] <- encoderArg{
-			series:     series,
-			dp:         dp,
-			unit:       unit,
-			annotation: annotation,
-			blockStart: dp.Timestamp.Truncate(blockSize),
+		// Distribute work.
+		workerEnqueue++
+		worker := workers[workerEnqueue%numWorkers]
+		worker.inputCh <- accumulateArg{
```

Comment:
Are there any perf implications from hammering channels so hard? Would it be worth trying to batch incoming points and sending that instead? (Probably not in this PR but for future work)

Reply:
Yeah we could do this as future work.

Reply:
I'll make a note.